compiler/daml-lf-ast/src/DA/Daml/LF/Ast/Numeric.hs

```diff
@@ -0,0 +1,105 @@
+-- the scale (0-37), the latter controlling how many digits appear after the
+-- decimal point. Furthermore, when reading or writing a numeric instance,
+-- we need to show every significant digit after the decimal point, in order
+-- to preserve the scale. For scale 0, the decimal point needs to be shown
+-- without a digit.
+--
+-- Internally we use Data.Decimal to represent these because it has the ability
+-- to encode every Numeric alongside its scale, and it mostly does what we want
+-- with Show and Read, with a few adjustments:
+--
+-- * we perform bounds checks with smart constructor 'numeric'
+-- * for scale 0, we have to handle the decimal point in the Show
+--   and Read instances manually
+-- * when reading, we check numeric bounds for scale and mantissa
+-- * we add Data, NFData, Generic instances
+-- * we don't add Num instances (for now anyway)
+--
+newtype Numeric = Numeric { numericDecimal :: Decimal }
+  deriving (Eq, Ord, Generic)
+
+-- | Smart constructor for Numeric literals.
+numeric :: Natural -> Integer -> Numeric
+numeric s m
+    | s > numericMaxScale = error "numeric error: scale too large"
+    | m > numericMaxMantissa = error "numeric error: mantissa too large"
+    | otherwise = Numeric $ Decimal (fromIntegral s) m
+
+-- | Upper bound for numeric scale (inclusive).
+numericMaxScale :: Natural
+numericMaxScale = 37
+
+-- | Upper bound for numeric mantissa (inclusive).
+numericMaxMantissa :: Integer
+numericMaxMantissa = 10^(38::Int)-1
+
+-- | Get scale associated with numeric literal. This is the
+-- number of decimal places after the decimal point. Ranges
+-- between 0 and 'numericMaxScale' (inclusive).
+numericScale :: Numeric -> Natural
+numericScale = fromIntegral . decimalPlaces . numericDecimal
+
+-- | Get mantissa associated with numeric literal. This is
+-- the raw integer value of the numeric before adding the
+-- decimal point. Ranges between 0 and 'numericMaxMantissa'
+-- (inclusive).
+numericMantissa :: Numeric -> Integer
+numericMantissa = decimalMantissa . numericDecimal
+
+instance Show Numeric where
+    showsPrec p n
+        | numericScale n == 0 = shows (numericDecimal n) . ("." ++)
+        | otherwise = showsPrec p (numericDecimal n)
+
+instance Read Numeric where
+    readsPrec p = mapMaybe postProcess . readsPrec p
+      where
+        postProcess :: (Decimal, String) -> Maybe (Numeric, String)
+        postProcess (d, xs) = do
+            let n = Numeric d
+            guard (numericValid n)
+            if numericScale n > 0 then
+                Just (n, xs)
+            else -- for scale == 0, we have to take the decimal point manually
+                case xs of
+                    '.':ys -> Just (n, ys)
+                    _ -> Nothing
+
+        numericValid :: Numeric -> Bool
+        numericValid n =
+            numericScale n <= numericMaxScale
+            && numericMantissa n <= numericMaxMantissa
+
+instance Data Numeric where
```

Comment:
Might be reasonable to propose the instance upstream.

Reply:
Yes, that's a good idea :)